PBD-37 - Add functionality to support looking up of docker hosts in AWS

eddgrant opened the pull request:
Introduces a new `AWSHostResolver` class which is responsible for looking up the private IP addresses of hosts based on a specified tag/ value. This functionality is mutually exclusive with the current behaviour of passing in a `docker_nodes` list. The new functionality is activated by the presence of a `AWS_DOCKER_HOST_TAG_VALUE` config item.

grahamhar commented on captain/tests/test_aws_host_resolver.py:
```diff
@@ -0,0 +1,162 @@
+class TestAWSHostResolver(unittest.TestCase):
+
+    def setUp(self):
+        self.ec2_resource = boto3.resource('ec2')
```
add `region_name='us-west-2'` to allow the tests to pass you probably have this set locally as an envvar

grahamhar commented on captain/aws_host_resolver.py:
```diff
@@ -0,0 +1,63 @@
+        self.cache_semaphore.acquire()
+        logger.debug(dict(Message="Locking semaphore acquired."))
+
+        if datetime.utcnow() > self.aws_cache_expiry_time:
```
do we cache?

eddgrant commented:
@grahamhar Managed to find a TTL based caching annotation which is more consistent with the existing LRUCache approach. Would you mind taking another look when you get a mo please?

grahamhar commented on captain/config.py:
```diff
@@ -14,9 +18,16 @@ def __init__(self):
-            raise Exception("SLUG_RUNNER_IMAGE should be specified")
+            raise Exception("SLUG_RUNNER_IMAGE must be specified")
+
+        self.aws_docker_host_tag_name = os.getenv("AWS_DOCKER_HOST_TAG_NAME", "role")
```
Could we move this section of checking up to line 10 to keep it near to the code setting nodes, I think it will make it cleaner to follow.

grahamhar commented on captain/connection.py:
```diff
@@ -15,17 +15,26 @@
 
 class Connection(object):
-    def __init__(self, config, verify=False):
+    def __init__(self, config, aws_host_resolver, verify=False):
```
could aws_host_resolver be None as a default if we are in a none AWS environment why would you pass an instance in?

grahamhar commented on captain_web.py:
```diff
@@ -1,30 +1,36 @@
 app.debug = True
 api = restful.Api(app, catch_all_404s=True)
 
+ec2_client = boto3.client('ec2')
```
I think this approach will break non AWS environments. In captain/config we raise an exception if a list of nodes is passed and aws config is attempted to be used. I think here we should only create the ec2_client/aws_host_resolver if the AWS tags config is present? 

This may also explain why there were issues with the tests passing, we maybe able to remove the fake credentials from the test config.

eddgrant commented:
@grahamhar Think I've addressed your comments, ended up using the factory pattern to separate AWS and non AWS behaviour. Possibly a little overkill but made it easy to test and felt like it helped the various responsibilities end up in the right place.  

Ready for re-review when you have a mo.

grahamhar commented on captain/aws.py:
```diff
@@ -0,0 +1,60 @@
+    Instances of this class are thread safe, hence a single instance can be used across many threads.
+    """
+
+    def __init__(self, ec2_client, dry_run=False, aws_call_interval_secs=10):
```
This default feels a bit low

grahamhar commented on captain/docker_node.py:
```diff
@@ -0,0 +1,87 @@
+        return ec2_client
+
+    def __create_aws_host_resolver(self, ec2_client):
+        aws_host_resolver = AWSHostResolver(ec2_client)
```
Should we be passing from config the aws_call_interval_secs value rather than using the default of 10 (If so ignore comment about low default).

eddgrant replied:
D'oh - quite right, I'd ommitted to pass the value in from the config. Have now done so, have left the default of 10 secs in the class itself as this could potentially be useful in generic cases. Have added a new config item with a default of 60 seconds. The config value is now used instead of the class level default. Do you think 60 secs is a reasonable default?
